File: ImperatorToCK3/Source/Imperator/Genes/AccessoryGene.cpp
#include "AccessoryGene.h"
#include "Log.h"
#include "ParserHelpers.h"
#include "CommonRegexes.h"



Imperator::AccessoryGene::AccessoryGene(std::istream& theStream)
{
	registerKeys();
	parseStream(theStream);
	clearRegisteredKeywords();
}


void Imperator::AccessoryGene::registerKeys() {
	registerKeyword("index", [this](std::istream& theStream) {
		index = commonItems::getInt(theStream);
	});
	registerKeyword("inheritable", [this](std::istream& theStream) {
		if (commonItems::getString(theStream) == "yes")
			inheritable = true;
	});
	registerRegex(commonItems::stringRegex, [this](const std::string& geneTemplateName, std::istream& theStream) {
		geneTemplates.insert({ geneTemplateName, AccessoryGeneTemplate(theStream) });
	});
	registerRegex(commonItems::catchallRegex, commonItems::ignoreItem);
}


const std::pair<const std::string, Imperator::AccessoryGeneTemplate>& Imperator::AccessoryGene::getGeneTemplateByIndex(const unsigned int indexInDna)
{
	for (auto& geneTemplateItr : geneTemplates) {
		if (geneTemplateItr.second.getIndex() == indexInDna)
			return geneTemplateItr;
	}
	LOG(LogLevel::Warning) << "Could not find gene template by index from DNA";
	return *geneTemplates.begin(); // fallback: return first element
}

File: ImperatorToCK3/Imperator/Genes/AccessoryGene.cs
using commonItems;
using System.Collections.Generic;
using System.Linq;

namespace ImperatorToCK3.Imperator.Genes {
	public class AccessoryGene : Parser {
		public uint Index { get; private set; } = 0;
		public PDXBool Inheritable { get; private set; } = new(false);
		public Dictionary<string, AccessoryGeneTemplate> GeneTemplates { get; private set; } = new();

		public AccessoryGene(BufferedReader reader) {
			RegisterKeys();
			ParseStream(reader);
			ClearRegisteredRules();
		}
		private void RegisterKeys() {
			RegisterKeyword("index", reader => {
				Index = (uint)reader.GetInt();
			});
			RegisterKeyword("inheritable", reader =>
				Inheritable = reader.GetPDXBool()
			);
			RegisterRegex(CommonRegexes.String, (reader, geneTemplateName) => {
				GeneTemplates.Add(geneTemplateName, new AccessoryGeneTemplate(reader));
			});
			RegisterRegex(CommonRegexes.Catchall, ParserHelpers.IgnoreItem);
		}
		public KeyValuePair<string, AccessoryGeneTemplate> GetGeneTemplateByIndex(uint indexInDna) {
			foreach (var geneTemplatePair in GeneTemplates) {
				if (geneTemplatePair.Value.Index == indexInDna) {
					return geneTemplatePair;
				}
			}
			Logger.Warn("Could not find gene template by index from DNA: " + indexInDna);
			return GeneTemplates.First(); // fallback: return first element
		}
	}
}
